rivertype/river_type.go

```diff
@@ -229,6 +230,49 @@ type AttemptError struct {
 	Trace string `json:"trace"`
 }
 
+type JobInsertParams struct {
+	Args         JobArgs
+	CreatedAt    *time.Time
+	EncodedArgs  []byte
+	Kind         string
+	MaxAttempts  int
+	Metadata     []byte
+	Priority     int
+	Queue        string
+	ScheduledAt  *time.Time
+	State        JobState
+	Tags         []string
+	UniqueKey    []byte
+	UniqueStates byte
+}
+
+// JobInsertMiddleware provides an interface for middleware that integrations can
+// use to encapsulate common logic around job insertion.
+//
+// Implementations should embed river.JobMiddlewareDefaults to inherit default
+// implementations for phases where no custom code is needed, and for forward
+// compatibility in case new functions are added to this interface.
+type JobInsertMiddleware interface {
+	// InsertMany is invoked around a batch insert operation. Implementations
+	// must always include a call to doInner to call down the middleware stack
+	// and perfom the batch insertion, and may run custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching insert
+	// operation, even if the inner insertion originally succeeded.
+	InsertMany(ctx context.Context, manyParams []*JobInsertParams, doInner func(context.Context) ([]*JobInsertResult, error)) ([]*JobInsertResult, error)
+}
+
+type WorkerMiddleware interface {
+	// Work is invoked after a job's JSON args being unmarshaled and before the
+	// job is worked. Implementations must always include a call to doInner to
+	// call down the middleware stack and perfom the batch insertion, and may run
+	// custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching work
+	// operation, even if the inner work originally succeeded.
+	Work(ctx context.Context, job *JobRow, doInner func(context.Context) error) error
```

Comment:
I just added a commit which splits the middleware interface in two. I'm not 100% settled on the design, but I was led here by trying to allow for middleware to be defined per-worker and related to this comment: https://github.com/riverqueue/river/pull/584#discussion_r1750869562

At insert time, we only have a `JobArgs` type and have no guaranteed access to the `Worker`. This inherently means that per-job-type job insertion middleware needs to be derived from the `JobArgs` and not connected to the `Worker`.

However when looking at `Work` middleware and its conceptual similarity to a worker's `Work` function, it can really only make sense for that to be dynamically defined on the stateful `Worker[T]` type. And if one were returning a general purpose `[]JobMiddleware` from their `JobArgs`, they would probably expect that the `Work()` functions on those would actually be used—but again they don't really make sense there rather than on the actual `Worker`.

As such, I've split the interface. While I don't think this split is ideal, I think it might be an unavoidable consequence of where we ended up on the ancient `JobArgs` vs `Worker` discussion from like a year ago 😃

On the plus side, along with this change, I added dynamic worker-level middleware defined by a `Middleware(job *Job[T]) []rivertype.WorkerMiddleware`. The executor calls into this via the work unit to extract the middleware that should be used for this specific job, and it applies those in order after the globally-defined `WorkerMiddleware` on the client config.

To make this work, I did have to move the arg deserialization to happen _before_ the middleware are called, but I think this is probably ok. This also adds some flexibility if there's a use case out there which demands access to the decoded args struct, because you can use that and the worker's other stateful properties when determining which middleware should be used. Not sure that's realistic or not 🤷‍♂️